fix: Specify an endpoint type for S3 VPC endpoint

haidaraM opened the pull request:
## Description
<!--- Describe your changes in detail -->
Correctly filter S3 endpoint based on the type of endpoint.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is now required because AWS seems to have added a new endpoint Interface for S3. If you don't filter the datasource for S3 endpoint on type, Terraform will fail because it will return multiple VPC endpoints
`Error: multiple VPC Endpoint Services matched; use additional constraints to reduce matches to a single VPC Endpoint Service`

Fix https://github.com/terraform-aws-modules/terraform-aws-vpc/issues/571

## Breaking Changes
<!-- Does this break backwards compatibility with the current major version? -->
<!-- If so, please provide an explanation why it is necessary. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

andormarkus commented:
Hi @haidaraM 

I have reviewed your PR I found the following issues:
- Documentation not updated
- Developer can create interface or gateway endpoint not both of them

However I dont see this one as new feature from AWS side. I think it is a bug.
Gateway is free, however interface cost money. I don't see a reason why anyone would use S3 interface instead of gateway.
If you got centralised vpc endpoints I don't see it the reson for it. 

`Gateway endpoints are free to provision, and there is not a strong use case for centralization.` [https://docs.aws.amazon.com/whitepapers/latest/building-scalable-secure-multi-vpc-network-infrastructure/centralized-access-to-vpc-private-endpoints.html](url)

Here is the difference between of the two: [https://gkzz.medium.com/what-is-the-differences-between-vpc-endpoint-gateway-endpoint-ae97bfab97d8](url)

discreet commented:
I am unable to make any VPC changes via Terraform and this module until this is fixed by adding the service_type

haidaraM commented:
Hi @andormarkus, Thanks for your feedback:
 - Just updated the documentation
 - It may be a bug and in fact I don't see a reason to use an Interface endpoint here but as it's supported by the API, I added a new variable to support both cases without changing the current behavior people expect

haidaraM commented:
Just saw this: https://github.com/terraform-aws-modules/terraform-aws-vpc/issues/445#issuecomment-643985043 
Maybe not a bug after all

andormarkus commented:
Hi @haidaraM 

I assume this is a new interface feature from AWS. I will close my PR #572 

However in this case this PR needs more modification:
- Current integration should only focus on gateway type S3 endpoint. Interface type endpoint should not be associated with routing tables
- It need a separate interface type integration for S3. Here is full commit example for RDS [https://github.com/terraform-aws-modules/terraform-aws-vpc/commit/5446e1c86a6341beef4c58445890c4b9fc96366f](url)
- Lot of documentation in the readme

antonbabenko commented:
[v2.69.0](https://github.com/terraform-aws-modules/terraform-aws-vpc/releases/v2.69.0) has been just released.

Hopefully, this is going to fix the issue AWS brought to us so unexpectedly.

I also put DynamoDB endpoint type as a configuration in case AWS decides to update it while I sleep :)

Thanks to everyone who worked on the solution for this!

bsakweson commented:
Here is how I resolved this issue at my end:

- Updated terraform `aws` provider minimum required version to `3.1.0`
- Change my `terraform-aws-vpc` to `v2.70.0`

I saw this error for the first time today and was able to fix it with those two updates.

artburkart commented:
I couldn't get it to work unless I used AWS Terraform provider `3.10.0`, because the "terraform-aws-vpc" module at version `2.70.0` wouldn't allow me to install an older version.

arik-porat commented:
Hi guys, how can I get the fix if I'm using terraform version 0.12.20?
This module latest version is only working for terraform version >= 0.12.21

haidaraM commented:
@arik-porat Update to at least Terraform 0.12.21 and update your AWS provider to at least 3.10

arik-porat commented:
Going from 0.12.20 to 0.12.21 just require to replace the terrafrom version or there's an upgrade procedure? didn't anything about migrating a minor terraform version.

haidaraM commented:
From the CHANGELOG, nothing special from 0.12.20 to 0.12.21: https://github.com/hashicorp/terraform/blob/v0.12/CHANGELOG.md#01221-february-19-2020. Just change the version and should work.

If your AWS provider is still on 2.X, you may need to do some changes: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-3-upgrade. 

Maybe do the updates and check terraform plan :-)

andormarkus commented:
Hi @haidaraM,

If you set S3 endpoint type to "Interface" it will cause an error as expected.
`Error: An Interface VPC Endpoint must always have at least one Security Group`

Fixed in #575

haidaraM commented:
Perfect @andormarkus 

Here is AWS news: https://aws.amazon.com/fr/about-aws/whats-new/2021/02/amazon-s3-now-supports-aws-privatelink/

andormarkus commented:
Hi @haidaraM 


Typical AWS 😄  They just releasing new features and updating documentation or making announcement later
